File: sdks/typescript/auth/src/store.ts
/**
 * Persistent credential store at the fixed, invocation-independent per-platform path:
 * `$XDG_CONFIG_HOME/oura-toolkit/` (→ `~/.config/oura-toolkit/`) on Unix/macOS,
 * `%LOCALAPPDATA%\oura-toolkit\` on Windows — the SAME directory and the SAME two JSON
 * records as the Rust companion (sdks/rust/oura-toolkit-auth/src/store.rs), so the CLI,
 * the MCP server, and this SDK companion all share one store.
 *
 * Two records live in the store dir:
 *
 * - **`credentials.json`** — {@link ClientCredentials} (the user's own OAuth app id +
 *   secret; BYO confidential client). Exists from `oura auth setup` onward.
 * - **`tokens.json`** — {@link Tokens} (access/refresh/expiry/scope). Exists only after
 *   a successful login and is rewritten on every refresh rotation.
 *
 * The on-disk field names are the Rust structs' serde output (`client_id`,
 * `client_secret`, `access_token`, `refresh_token`, `expires_at`, optional `scope` /
 * `token_type` OMITTED when absent) — pinned by a fixture test in `tests/store.test.cjs`.
 *
 * File hygiene: records are written atomically (uniquely named temp file in the same
 * directory + rename) and, on Unix, chmod'd `0600` with a `0700` parent dir. On Windows
 * those modes are no-ops — protection relies on `%LOCALAPPDATA%`'s user-private ACLs.
 *
 * Secret hygiene: {@link ClientCredentials} and {@link Tokens} REDACT their secret
 * fields from `JSON.stringify`, `util.inspect` (and therefore `console.log`), and
 * `toString`. The store's own serializer is the only path that writes the real values —
 * a stray log statement can never leak a secret, and `JSON.stringify(tokens)` is
 * deliberately NOT a way to persist tokens.
 */

import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";

import { NoConfigDirError, StoreFormatError } from "./errors";

const INSPECT = Symbol.for("nodejs.util.inspect.custom");
const REDACTED = "[REDACTED]";

/**
 * The user's own Oura OAuth application credentials (BYO confidential client).
 *
 * The secret lives in a private field: it is invisible to `JSON.stringify`, spreads,
 * `Object.entries`, and `util.inspect` — read it via {@link clientSecret}. That accessor
 * is a METHOD, not a getter, deliberately: `util.inspect(x, { getters: true })` evaluates
 * accessor properties (and would print the secret), but never invokes plain methods.
 */
export class ClientCredentials {
  readonly clientId: string;
  readonly #clientSecret: string;

  constructor(args: { clientId: string; clientSecret: string }) {
    this.clientId = args.clientId;
    this.#clientSecret = args.clientSecret;
  }

  /** The client secret. Never appears in logs/serialization — see class docs. */
  clientSecret(): string {
    return this.#clientSecret;
  }

  /** Redacted: `JSON.stringify` must never emit the secret. Persist via {@link TokenStore}. */
  toJSON(): unknown {
    return { client_id: this.clientId, client_secret: REDACTED };
  }

  toString(): string {
    return `ClientCredentials { clientId: ${JSON.stringify(this.clientId)}, clientSecret: ${REDACTED} }`;
  }

  /** Redacted `util.inspect` / `console.log` rendering. */
  [INSPECT](): string {
    return this.toString();
  }
}

/**
 * The persisted OAuth token set. Client credentials live in their own record
 * ({@link ClientCredentials}), not here.
 *
 * The access and refresh tokens live in private fields and are redacted from
 * `JSON.stringify` / `util.inspect` / `toString` (see module docs). Their accessors are
 * METHODS, not getters — `util.inspect(x, { getters: true })` evaluates getters but
 * never invokes plain methods.
 */
export class Tokens {
  readonly #accessToken: string;
  readonly #refreshToken: string;
  /** Absolute expiry as a Unix timestamp (seconds). */
  readonly expiresAt: number;
  readonly scope?: string;
  readonly tokenType?: string;

  constructor(args: {
    accessToken: string;
    refreshToken: string;
    expiresAt: number;
    scope?: string;
    tokenType?: string;
  }) {
    this.#accessToken = args.accessToken;
    this.#refreshToken = args.refreshToken;
    this.expiresAt = args.expiresAt;
    this.scope = args.scope;
    this.tokenType = args.tokenType;
  }

  accessToken(): string {
    return this.#accessToken;
  }

  /**
   * Oura ROTATES this on every refresh and invalidates the previous value — whoever
   * refreshes must persist the newly returned token or the next refresh 400s.
   */
  refreshToken(): string {
    return this.#refreshToken;
  }

  /** True if the access token is expired (or within `skewSecs` of expiring). */
  isExpired(skewSecs: number): boolean {
    return Math.floor(Date.now() / 1000) + skewSecs >= this.expiresAt;
  }

  /** Redacted: `JSON.stringify` must never emit tokens. Persist via {@link TokenStore}. */
  toJSON(): unknown {
    return {
      access_token: REDACTED,
      refresh_token: REDACTED,
      expires_at: this.expiresAt,
      scope: this.scope,
      token_type: this.tokenType,
    };
  }

  toString(): string {
    return (
      `Tokens { accessToken: ${REDACTED}, refreshToken: ${REDACTED}, ` +
      `expiresAt: ${this.expiresAt}, scope: ${JSON.stringify(this.scope)}, ` +
      `tokenType: ${JSON.stringify(this.tokenType)} }`
    );
  }

  /** Redacted `util.inspect` / `console.log` rendering. */
  [INSPECT](): string {
    return this.toString();
  }
}

/** The locked config-directory name (CLAUDE.md → NAMING), identical on every platform. */
export const APP_DIR_NAME = "oura-toolkit";

/** Injectable environment lookup (tests never mutate `process.env`). */
export type EnvLookup = (key: string) => string | undefined;

/**
 * Resolve the fixed, invocation-independent config dir from an injected env lookup:
 *
 * - **Unix (incl. macOS):** `$XDG_CONFIG_HOME/oura-toolkit`, falling back to
 *   `$HOME/.config/oura-toolkit` (locked decision — deliberately NOT
 *   `~/Library/Application Support`).
 * - **Windows:** `%LOCALAPPDATA%\oura-toolkit` — Local, NOT Roaming: roaming profiles
 *   sync `%APPDATA%` off the machine, which would copy plaintext secrets to file
 *   servers/backups.
 *
 * EMPTY and RELATIVE env values are treated as absent (XDG spec): a relative base would
 * make where secrets land depend on the process cwd, breaking invocation independence
 * (the path MUST be identical under `npx`, `bunx`, and a brew binary).
 */
export function configDirFrom(
  env: EnvLookup,
  platform: NodeJS.Platform = process.platform
): string {
  const p = platform === "win32" ? path.win32 : path.posix;
  const usable = (key: string): string | undefined => {
    const value = env(key);
    return value && p.isAbsolute(value) ? value : undefined;
  };

  if (platform === "win32") {
    const base = usable("LOCALAPPDATA");
    if (base !== undefined) return p.join(base, APP_DIR_NAME);
    throw new NoConfigDirError("%LOCALAPPDATA% unset or not an absolute path");
  }

  const xdg = usable("XDG_CONFIG_HOME");
  if (xdg !== undefined) return p.join(xdg, APP_DIR_NAME);
  const home = usable("HOME");
  if (home !== undefined) return p.join(home, ".config", APP_DIR_NAME);
  throw new NoConfigDirError("$XDG_CONFIG_HOME / $HOME unset or not an absolute path");
}

/** Handle to the on-disk store directory. */
export class TokenStore {
  readonly dir: string;

  /** Store rooted at an explicit directory (used by tests). */
  constructor(dir: string) {
    this.dir = dir;
  }

  /** Store at the default per-platform config location (see {@link configDirFrom}). */
  static default(env: EnvLookup = (key) => process.env[key]): TokenStore {
    return new TokenStore(configDirFrom(env));
  }

  /** Path of the client-credentials record. */
  credentialsPath(): string {
    return path.join(this.dir, "credentials.json");
  }

  /** Path of the token record. */
  tokensPath(): string {
    return path.join(this.dir, "tokens.json");
  }

  /** Load the client credentials, or `null` if `auth setup` has never run. */
  loadCredentials(): ClientCredentials | null {
    const record = readRecord(this.credentialsPath());
    if (record === null) return null;
    return new ClientCredentials({
      clientId: requireString(record, "client_id"),
      clientSecret: requireString(record, "client_secret"),
    });
  }

  /** Persist the client credentials (`0600`, atomic). */
  saveCredentials(credentials: ClientCredentials): void {
    // Explicit plain object — deliberately NOT JSON.stringify(credentials), whose
    // toJSON() redacts. Field names are the Rust structs' serde output (fixture-pinned).
    const record = {
      client_id: credentials.clientId,
      client_secret: credentials.clientSecret(),
    };
    this.#writeRecord(this.credentialsPath(), record);
  }

  /** Load the tokens, or `null` if no login has succeeded yet. */
  loadTokens(): Tokens | null {
    const record = readRecord(this.tokensPath());
    if (record === null) return null;
    return new Tokens({
      accessToken: requireString(record, "access_token"),
      refreshToken: requireString(record, "refresh_token"),
      expiresAt: requireNumber(record, "expires_at"),
      scope: optionalString(record, "scope"),
      tokenType: optionalString(record, "token_type"),
    });
  }

  /**
   * Persist the tokens (`0600`, atomic). Callers refreshing MUST persist the rotated
   * refresh token (Oura invalidates the previous one) — {@link TokenManager} does.
   */
  saveTokens(tokens: Tokens): void {
    // Optionals are OMITTED when absent, matching serde's skip_serializing_if.
    const record: Record<string, string | number> = {
      access_token: tokens.accessToken(),
      refresh_token: tokens.refreshToken(),
      expires_at: tokens.expiresAt,
    };
    if (tokens.scope !== undefined) record.scope = tokens.scope;
    if (tokens.tokenType !== undefined) record.token_type = tokens.tokenType;
    this.#writeRecord(this.tokensPath(), record);
  }

  #writeRecord(filePath: string, record: object): void {
    this.#ensureDir();
    writeSecure(filePath, JSON.stringify(record, null, 2));
  }

  #ensureDir(): void {
    fs.mkdirSync(this.dir, { recursive: true, mode: 0o700 });
    if (process.platform !== "win32") {
      // mkdirSync's mode only applies on creation; enforce 0700 on the existing dir too.
      fs.chmodSync(this.dir, 0o700);
    }
  }
}

function readRecord(filePath: string): Record<string, unknown> | null {
  let raw: string;
  try {
    raw = fs.readFileSync(filePath, "utf8");
  } catch (e) {
    const code = (e as NodeJS.ErrnoException).code;
    if (code === "ENOENT") return null;
    // A directory (or other non-file) where a record should be is a malformed store, not
    // an opaque OS error — surface it as the typed StoreFormatError callers already handle.
    if (code === "EISDIR") {
      throw new StoreFormatError(`${filePath} is a directory, not a token-store file`);
    }
    throw e;
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (e) {
    throw new StoreFormatError(`${filePath} is not valid JSON: ${(e as Error).message}`);
  }
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new StoreFormatError(`${filePath} must contain a JSON object`);
  }
  return parsed as Record<string, unknown>;
}

function requireString(record: Record<string, unknown>, field: string): string {
  const value = record[field];
  if (typeof value !== "string") {
    throw new StoreFormatError(`missing or non-string field "${field}"`);
  }
  return value;
}

function requireNumber(record: Record<string, unknown>, field: string): number {
  const value = record[field];
  if (typeof value !== "number") {
    throw new StoreFormatError(`missing or non-number field "${field}"`);
  }
  return value;
}

function optionalString(record: Record<string, unknown>, field: string): string | undefined {
  const value = record[field];
  if (value === undefined || value === null) return undefined;
  if (typeof value !== "string") {
    throw new StoreFormatError(`non-string field "${field}"`);
  }
  return value;
}

/**
 * Atomic write with owner-only perms: write a UNIQUELY named temp file in the same
 * directory, fsync, rename into place. The unique name means two concurrent writers can
 * never truncate each other's in-flight temp file; the atomic rename makes the outcome
 * last-writer-wins, never a corrupt mix. On Windows the chmod is a no-op (ACLs protect).
 */
function writeSecure(filePath: string, data: string): void {
  const dir = path.dirname(filePath);
  const tmp = path.join(dir, `.tmp-${crypto.randomBytes(8).toString("hex")}`);
  const fd = fs.openSync(tmp, "wx", 0o600);
  try {
    fs.writeSync(fd, data);
    fs.fsyncSync(fd);
  } finally {
    fs.closeSync(fd);
  }
  try {
    fs.renameSync(tmp, filePath);
  } catch (e) {
    fs.rmSync(tmp, { force: true });
    throw e;
  }
  if (process.platform !== "win32") {
    // openSync's 0600 is umask-masked at creation; pin the final mode exactly.
    fs.chmodSync(filePath, 0o600);
  }
}

File: sdks/go/auth/store.go
package auth

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

// appDirName is the locked config-directory name (CLAUDE.md → NAMING), identical under
// npx/bunx/brew and identical to the Rust companion's store.
const appDirName = "oura-toolkit"

// ClientCredentials is the user's own Oura OAuth application id + secret (BYO
// confidential client). JSON field names are pinned to the Rust companion's serde output
// (credentials.json is shared with the CLI) — see the fixture test.
//
// String/GoString REDACT the secret so a stray %v/%+v/%#v can never leak it into logs.
type ClientCredentials struct {
	ClientID     string `json:"client_id"`
	ClientSecret string `json:"client_secret"`
}

func (c ClientCredentials) String() string {
	return fmt.Sprintf("ClientCredentials{ClientID:%q ClientSecret:[REDACTED]}", c.ClientID)
}

// GoString redacts under %#v (fmt uses GoStringer there, not Stringer).
func (c ClientCredentials) GoString() string { return c.String() }

// Tokens is the persisted OAuth token set. Client credentials live in their own record
// (ClientCredentials), not here. JSON field names are pinned to the Rust companion's
// serde output (tokens.json is shared with the CLI) — see the fixture test.
//
// String/GoString REDACT the token fields so a stray %v/%+v/%#v can never leak them.
type Tokens struct {
	AccessToken string `json:"access_token"`
	// RefreshToken is rotated by Oura on every refresh, invalidating the previous value —
	// always persist the newly returned one or the next refresh 400s.
	RefreshToken string `json:"refresh_token"`
	// ExpiresAt is the absolute expiry as a Unix timestamp (seconds).
	ExpiresAt int64  `json:"expires_at"`
	Scope     string `json:"scope,omitempty"`
	TokenType string `json:"token_type,omitempty"`
}

func (t Tokens) String() string {
	return fmt.Sprintf(
		"Tokens{AccessToken:[REDACTED] RefreshToken:[REDACTED] ExpiresAt:%d Scope:%q TokenType:%q}",
		t.ExpiresAt, t.Scope, t.TokenType,
	)
}

// GoString redacts under %#v (fmt uses GoStringer there, not Stringer).
func (t Tokens) GoString() string { return t.String() }

// expired reports whether the access token is expired, or within skew of expiring.
func (t *Tokens) expired(skew time.Duration) bool {
	return time.Now().Add(skew).Unix() >= t.ExpiresAt
}

// Store is a handle to the on-disk store directory. Two records live in it:
//
//   - credentials.json — ClientCredentials; exists from `oura auth setup` onward.
//   - tokens.json — Tokens; exists after a successful login, rewritten on every rotation.
//
// Separating the two means a failed or abandoned login never loses the pasted secret.
// Everything is written via an atomic, uniquely named temp-file + rename with 0600 perms
// on Unix (Windows relies on %LOCALAPPDATA%'s user-private ACLs).
type Store struct {
	dir string
}

// NewStore opens the store at the default per-platform config location
// ($XDG_CONFIG_HOME/oura-toolkit → ~/.config/oura-toolkit on Unix/macOS,
// %LOCALAPPDATA%\oura-toolkit on Windows).
func NewStore() (*Store, error) {
	dir, err := configDirFrom(runtime.GOOS, os.LookupEnv)
	if err != nil {
		return nil, err
	}
	return &Store{dir: dir}, nil
}

// NewStoreAt opens a store rooted at an explicit directory (used by tests).
func NewStoreAt(dir string) *Store {
	return &Store{dir: dir}
}

// Dir returns the store directory.
func (s *Store) Dir() string { return s.dir }

// CredentialsPath is the client-credentials record path.
func (s *Store) CredentialsPath() string { return filepath.Join(s.dir, "credentials.json") }

// TokensPath is the token record path.
func (s *Store) TokensPath() string { return filepath.Join(s.dir, "tokens.json") }

// LoadCredentials returns the client credentials, or (nil, nil) if `oura auth setup` has
// never run.
func (s *Store) LoadCredentials() (*ClientCredentials, error) {
	var c ClientCredentials
	ok, err := s.loadJSON(s.CredentialsPath(), &c)
	if err != nil || !ok {
		return nil, err
	}
	return &c, nil
}

// SaveCredentials persists the client credentials (0600, atomic).
func (s *Store) SaveCredentials(c *ClientCredentials) error {
	return s.saveJSON(s.CredentialsPath(), c)
}

// LoadTokens returns the tokens, or (nil, nil) if no login has succeeded yet.
func (s *Store) LoadTokens() (*Tokens, error) {
	var t Tokens
	ok, err := s.loadJSON(s.TokensPath(), &t)
	if err != nil || !ok {
		return nil, err
	}
	return &t, nil
}

// SaveTokens persists the tokens (0600, atomic). Callers refreshing MUST persist the
// rotated refresh token (Oura invalidates the previous one), and anything that refreshes
// and rewrites tokens must do so under LockExclusive — as Manager does — or it can burn a
// rotation another process just persisted.
func (s *Store) SaveTokens(t *Tokens) error {
	return s.saveJSON(s.TokensPath(), t)
}

func (s *Store) loadJSON(path string, into any) (bool, error) {
	data, err := os.ReadFile(path)
	if errors.Is(err, fs.ErrNotExist) {
		return false, nil
	}
	if err != nil {
		return false, fmt.Errorf("token store i/o error: %w", err)
	}
	if err := json.Unmarshal(data, into); err != nil {
		return false, fmt.Errorf("token store format error: %w", err)
	}
	return true, nil
}

func (s *Store) saveJSON(path string, v any) error {
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return fmt.Errorf("token store format error: %w", err)
	}
	if err := s.ensureDir(); err != nil {
		return err
	}
	return writeSecure(path, data)
}

// StoreLock is an exclusive advisory lock on the store, released by Unlock (or process
// exit). Hold it across a reload → refresh → persist critical section (see Manager).
type StoreLock struct {
	f *os.File
}

// Unlock releases the lock and closes the underlying file.
func (l *StoreLock) Unlock() error {
	err := unlockFile(l.f)
	if cerr := l.f.Close(); err == nil {
		err = cerr
	}
	return err
}

// LockExclusive takes a BLOCKING exclusive lock on the store's .lock file — flock on
// Unix, LockFileEx on Windows: the same primitives as Rust std's File::lock, so this
// coordinates with the CLI and MCP server sharing the store.
//
// The lock is cooperative (it excludes only other holders of this protocol, not
// arbitrary writers), and mutual exclusion rests on the .lock file's identity — deleting
// it while held lets the next locker acquire a fresh one and defeats coordination.
func (s *Store) LockExclusive() (*StoreLock, error) {
	f, err := s.openLockFile()
	if err != nil {
		return nil, err
	}
	if err := lockFile(f); err != nil {
		f.Close()
		return nil, fmt.Errorf("token store lock error: %w", err)
	}
	return &StoreLock{f: f}, nil
}

// TryLockExclusive is the non-blocking variant: (nil, nil) means another process
// currently holds the lock.
func (s *Store) TryLockExclusive() (*StoreLock, error) {
	f, err := s.openLockFile()
	if err != nil {
		return nil, err
	}
	ok, err := tryLockFile(f)
	if err != nil {
		f.Close()
		return nil, fmt.Errorf("token store lock error: %w", err)
	}
	if !ok {
		f.Close()
		return nil, nil
	}
	return &StoreLock{f: f}, nil
}

func (s *Store) openLockFile() (*os.File, error) {
	if err := s.ensureDir(); err != nil {
		return nil, err
	}
	f, err := os.OpenFile(filepath.Join(s.dir, ".lock"), os.O_CREATE|os.O_RDWR, 0o600)
	if err != nil {
		return nil, fmt.Errorf("token store i/o error: %w", err)
	}
	return f, nil
}

func (s *Store) ensureDir() error {
	if err := os.MkdirAll(s.dir, 0o700); err != nil {
		return fmt.Errorf("token store i/o error: %w", err)
	}
	return setDirPrivate(s.dir)
}

// configDirFrom resolves the fixed, invocation-independent config dir from an injected
// env lookup (testable on every OS without racy os.Setenv):
//
//   - Unix (incl. macOS): $XDG_CONFIG_HOME/oura-toolkit, falling back to
//     $HOME/.config/oura-toolkit. The XDG path is a locked decision (CLAUDE.md) —
//     deliberately NOT macOS's ~/Library/Application Support.
//   - Windows: %LOCALAPPDATA%\oura-toolkit. Local, NOT roaming: roaming profiles sync
//     %APPDATA% to file servers/backups at logoff, which would copy plaintext secrets off
//     the machine.
//
// Empty and RELATIVE values are treated as absent (XDG spec: relative values should be
// ignored) — a relative base would make where secrets land depend on the process cwd,
// breaking the invocation-independence invariant.
func configDirFrom(goos string, lookup func(string) (string, bool)) (string, error) {
	usable := func(key string) (string, bool) {
		v, ok := lookup(key)
		if !ok || v == "" || !isAbs(goos, v) {
			return "", false
		}
		return v, true
	}

	if goos == "windows" {
		if base, ok := usable("LOCALAPPDATA"); ok {
			return filepath.Join(base, appDirName), nil
		}
		return "", fmt.Errorf("%w (%%LOCALAPPDATA%% unset or not an absolute path)", ErrNoConfigDir)
	}

	if xdg, ok := usable("XDG_CONFIG_HOME"); ok {
		return filepath.Join(xdg, appDirName), nil
	}
	if home, ok := usable("HOME"); ok {
		return filepath.Join(home, ".config", appDirName), nil
	}
	return "", fmt.Errorf("%w ($XDG_CONFIG_HOME / $HOME unset or not an absolute path)", ErrNoConfigDir)
}

// isAbs is goos-parameterized (unlike filepath.IsAbs, which is compile-time) so both
// platform branches of configDirFrom are unit-tested on every CI leg.
func isAbs(goos, path string) bool {
	if goos == "windows" {
		// Drive-rooted (C:\ or C:/) or UNC (\\server\share).
		if len(path) >= 3 && path[1] == ':' && (path[2] == '\\' || path[2] == '/') {
			return true
		}
		return strings.HasPrefix(path, `\\`)
	}
	return strings.HasPrefix(path, "/")
}

// writeSecure writes atomically with owner-only perms: a UNIQUELY named temp file in the
// same directory (os.CreateTemp creates 0600), fsync, rename into place. The unique name
// means two concurrent writers can never truncate each other's in-flight temp file; the
// atomic rename makes the outcome last-writer-wins, never a corrupt mix.
func writeSecure(path string, data []byte) error {
	dir := filepath.Dir(path)
	tmp, err := os.CreateTemp(dir, ".tmp-*")
	if err != nil {
		return fmt.Errorf("token store i/o error: %w", err)
	}
	tmpName := tmp.Name()
	cleanup := func(err error) error {
		tmp.Close()
		os.Remove(tmpName)
		return fmt.Errorf("token store i/o error: %w", err)
	}
	if _, err := tmp.Write(data); err != nil {
		return cleanup(err)
	}
	if err := tmp.Sync(); err != nil {
		return cleanup(err)
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpName)
		return fmt.Errorf("token store i/o error: %w", err)
	}
	if err := os.Rename(tmpName, path); err != nil {
		os.Remove(tmpName)
		return fmt.Errorf("token store i/o error: %w", err)
	}
	return setFilePrivate(path)
}
